apis/accounts/auth/login.yaml

```diff
@@ -0,0 +1,55 @@
+get:
+  tags:
+    - AccountAuth
+  operationId: "LoginStatus"
+  summary: "Check login status."
+  description: |
+    Convenience method to check if user has signed up, and their token is valid.
+  security:
+    - cookieAuth: []
+  responses:
+    "200":
+      content:
+        application/json:
+          schema:
+            title: LoginStatusResponse
+            type: object
+            properties:
+              has_signed_up:
+                type: boolean
+    "500":
+      $ref: "../../../validator-oapi.yaml#/components/responses/InternalError"
+
+post:
+  tags:
+    - AccountAuth
+  operationId: "Login"
+  summary: "Login"
+  description: |
+    MUST validate the provided passwords against a local “Record”. After successful validation MUST return a Set-Cookie header
+  requestBody:
+    content:
+      application/json:
+        schema:
+          type: object
+          properties:
+            password:
+              type: string
+            password_confirmation:
+              type: string
+  responses:
+    "200":
+      description: |
+        Successfully authenticated. The session ID is returned in a cookie named `SESSION`. You need to include this cookie in subsequent requests.
+        - HttpOnly: MUST be included
```

Comment:
Is it possible to avoid cookies? Because as soon as you introduce cookies for authentication server become stateful. Also because `Max-Age` is not specified its possible to create cookies which will never be deleted, and so can be used if cookie become stolen. Is it possible to use token authentication insteed?

Reply:
Could you describe in more detail what you refer to with "token authentication"?

Given that each node will likely implement different flavours of authentication, using HttpOnly cookies would simplify the UI implementation. The browser will just re-send whatever that specific node has sent in the Set-Cookie header.

- You can make the cookie stateless by signing it with a known key and adding data to it.
- You can also add a timestamp to the cookie's content and use that to invalidate it in the backend, irrespective of the Max-Age directive

Reply:
You can check for Json Web Token about token authentication which could be passed via headers or request string. The biggest difference is that tokens will not be managed by browsers.
This feature protects validators from loosing cookies through malicious browser extensions or malware.
As soon as communication finishes, tokens become not valid anymore.

Reply:
Lighthouse has an auth scheme in its existing VC API:

- A secret is passed by the `Authorization` HTTP header: https://lighthouse-book.sigmaprime.io/api-vc-auth-header.html
- Since the secret is a secp256k1 public key, the server also includes a `Signature` header in all responses (note, this is not safe from replay attacks).

Here's an example of a `curl` request using this scheme:

```bash
curl -v localhost:5062/lighthouse/version -H "Authorization: Basic api-token-0x03eace4c98e8f77477bb99efb74f9af10d800bd3318f92c33b719a4644254d4123"
```

This approach is simpler for the VC, but it doesn't allow for user specified passwords. Our design ethos is to keep the VC HTTP server as simple as possible and to handle all the user-facing authentication in the GUI. Here's a design doc we drew up a while ago: https://hackmd.io/zS3h-Kk_TTmYhvyoHap6DQ

Basically, we expect the VC and BN to be hiding behind a reverse-proxy provided by the GUI web app. This avoids needing to provide HTTPS certificates for the BN, VC *and* web application.

Reply:
@paulhauner This Signature Header is very cool idea! Is this like a roll-your-own TLS strategy to guarantee integrity of the responses? If so, could we take it a step further and make the client encrypt the requests with the pubkey so that the keystores + password are not sent in plain text? https://lighthouse-book.sigmaprime.io/api-vc-sig-header.html

Reply:
>  If so, could we take it a step further and make the client encrypt the requests with the pubkey so that the keystores + password are not sent in plain text?

As I understand it, encrypting the body of a HTTP request is generally not done. That's because the path, headers, etc still remain plain-text and it does not prevent replay attacks. Usually transport-level security (HTTPS) is used to fully solve all these issues.

Encrypting *all* the request bodies seems overkill to me, it would break the ability to do a simple `curl` request. However, perhaps encrypting passwords with that key is a useful last-line-of-defense. I'm unsure.

I am cautious to provide a false sense of security when using HTTP. I was even unsure about adding the `Signature` header, but I went ahead since it's fully optional on the client side.

Reply:
The beacon client should not provide extensive user authentication & authorization. A simple API key mechanism is enough for putting a UI-based user authenticatoin & authorization on top. In the case of implementing authentication & authorization on the beacon client, where do you draw the line with features? What about 2FA? Different storage techniques for sensible data? Etc.

Handling a JWT (Json Web Token) only makes sense when complex user management is needed.

When HTTPS is needed because access is provided on an unsecure network, a simple SSL termination service (e. g. nginx, haproxy, ...) would suffice and is more secure than handling SSL on the beacon client itself.